Implementation of the cumulative distribution function and its inverse

vishwakftw opened the pull request:
This PR adds the `.cdf` and `.icdf` methods for distributions. @alicanb implemented these in the `Distribution` base class.

Furthermore, this PR adds these methods for `Exponential`, `Normal`, `Cauchy` and for `TransformedDistribution`. `Pareto` and `Gumbel` are converted to `TransformedDistribution` with base classes as `Exponential` and `Uniform` respectively.

Tests for checking the correctness of `.cdf` and `.icdf` methods have also been added. An autograd based check is added to check the correctness of `.cdf` and `.log_prob` as well (this doesn't depend on `TEST_NUMPY`).

Joint work with @fritzo and @alicanb . Checked and approved by @fritzo .

alicanb committed:
add cdf and icdf to normal

vishwakftw committed:
New CDF and ICDF implementations
1. Cauchy
2. Exponential
3. Laplace (Only CDF)
4. Pareto

vishwakftw committed:
Major: Add .cdf and .icdf methods for TransformedDistributions
Minor:
1. Convert Pareto and Gumbel to TransformedDistribution
2. Add .cdf and .icdf for Uniform
3. Temporarily remove .cdf from Laplace

vishwakftw committed:
Add SciPy / NumPy independent tests for cdf and icdf invertibility

vishwakftw committed:
Addition of NumPy / SciPy independent tests for .log_prob using derivative of .cdf

vishwakftw committed:
Merge branch 'master' into cdf-icdf-methods

vishwakftw committed:
Bug fixes
1. Fix the size issue with Gumbel as a transformed distribution
2. Add the scalar params test for Gumbel

vishwakftw committed:
Remove batch_shape calculation after making Pareto distribution a Transformed Distribution

fritzo committed:
Fix out-of-range test parameters

vishwakftw committed:
Add .cdf and .icdf for Laplace distribution

fritzo reviewed (approved):
LGTM Previous review: https://github.com/probtorch/pytorch/pull/122

apaszke commented on test/test_distributions.py:
```diff
@@ -1161,6 +1161,47 @@ def test_beta_sample(self):
+                dist = Dist(**param)
+                samples = dist.sample()
+                if not samples.requires_grad:
+                    continue
```
Why would an output of `sample` ever require grad?

fritzo replied:
Good point, I think this should be
```py
samples = Variable(samples.data, requires_grad=True)
```
to make `grad()` work below.

apaszke commented on test/test_distributions.py:
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
     def test_mean(self):
         for pytorch_dist, scipy_dist in self.distribution_pairs:
+            if isinstance(pytorch_dist, Cauchy):
+                continue
```
A comment would be welcome (I have no idea why do we skip Cauchy)

apaszke commented on test/test_distributions.py:
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
 
     def test_variance_stddev(self):
         for pytorch_dist, scipy_dist in self.distribution_pairs:
+            if isinstance(pytorch_dist, Cauchy):
```
Same here

apaszke commented on test/test_distributions.py:
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
                 self.assertEqual(pytorch_dist.stddev, scipy_dist.var() ** 0.5, message=pytorch_dist)
 
+    def test_cdf(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
```
Is it really one of those tests? `cdf` is deterministic, so it shouldn't really matter, right? Or is it because of numerical instabilities?

alicanb replied:
I wanted to fix the seed just in case because of sampling, Note there is unnecessary though...

apaszke commented on torch/distributions/cauchy.py:
```diff
@@ -53,5 +53,13 @@ def log_prob(self, value):
+        return torch.atan((value - self.loc) / self.scale) / math.pi + 0.5
+
+    def icdf(self, value):
+        self._validate_log_prob_arg(value)
```
we might want to rename this to `_validate_sample`, since it's no longer used for args of log prob only

vishwakftw committed:
Address comments in the PR
1. Add comment regarding skipping Cauchy test in mean and stddev
2. Remove setting the random seed in cdf and icdf tests
3. Modify the test_cdf_log_prob, such that samples.requires_grad = True for computing gradient
